// @openapi-client/adapter-fetch@0.2.1/build/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function(mod) {
    return (mod && mod.__esModule) ? mod : {
        "default": mod
    };
};
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.FetchAdapter = void 0;
const querystring_1 = __importDefault(require("querystring"));
class FetchAdapter {
    constructor(options) {
        this.options = {};
        this.options = options;
    }
    async request(method, path, body, query, options) {
        const mergeOptions = { ...this.options,
            ...options
        };
        const queryString = query ?
            '?' + querystring_1.default.stringify(query) :
            '';
        try {
            const response = await fetch(`${mergeOptions.baseURL || ''}${path}${queryString}`, {
                method,
                body: body ? JSON.stringify(body) : null,
                ...mergeOptions,
            });
            return {
                status: response.status,
                statusText: response.statusText,
                data: await response.json(),
                headers: [...response.headers.entries()].reduce((acc, [k, v]) => {
                    acc[k] = v;
                    return acc;
                }, {}),
            };
        } catch (err) {
            err.isOpenApiError = true;
            throw err;
        }
    }
    withOptions(options) {
        return new FetchAdapter(options);
    }
}
exports.FetchAdapter = FetchAdapter;
