// @smartsupp/smartsupp-message@0.10.1/lib/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ChannelType = exports.MessageSubType = exports.MessageType = exports.AttachmentInvalidReason = exports.AttachmentType = exports.MessageActionType = void 0;
exports.MessageActionType = {
    Text: 'text',
    OpenUrl: 'open_url',
};
exports.AttachmentType = {
    Card: 'card',
    Cards: 'cards',
    File: 'file',
    Image: 'image',
    Media: 'media',
    Invalid: 'invalid',
};
exports.AttachmentInvalidReason = {
    InvalidSize: 'invalid_size',
    InvalidType: 'invalid_type',
};
exports.MessageType = {
    Message: 'message',
    Event: 'event',
    Help: 'help',
};
exports.MessageSubType = {
    Agent: 'agent',
    Contact: 'contact',
    Trigger: 'trigger',
    Bot: 'bot',
    System: 'system',
    AgentExternal: 'agent_external',
    FoundEmail: 'found_email',
    ChatResolve: 'chat_resolve',
};
exports.ChannelType = {
    Default: 'default',
    Email: 'email',
    FacebookMessenger: 'facebook_messenger',
    Smartsupp: 'smartsupp',
};
